pkg/agent/controller/packetsampling/packetin.go

```diff
@@ -0,0 +1,162 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packetsampling
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/gopacket"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	"antrea.io/libOpenflow/protocol"
+	"antrea.io/libOpenflow/util"
+	"antrea.io/ofnet/ofctrl"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+)
+
+func (c *Controller) HandlePacketIn(pktIn *ofctrl.PacketIn) error {
+	klog.V(4).Infof("PacketIn for PacketSampling: %+v", pktIn.PacketIn)
+	if !c.packetSamplingSynced() {
+		return errors.New("PacketSampling controller is not started")
+	}
+	oldPS, samplingState, shouldSkip, err := c.parsePacketIn(pktIn)
+	if err != nil {
+		return fmt.Errorf("parsePacketIn error: %v", err)
+	}
+	if shouldSkip {
+		return nil
+	}
+
+	// Retry when update CRD conflict which caused by multiple agents updating one CRD at same time.
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ps, err := c.packetSamplingInformer.Lister().Get(oldPS.Name)
+		if err != nil {
+			return fmt.Errorf("get PacketSampling failed: %w", err)
+		}
+
+		if samplingState != nil {
+			shouldUpdate := samplingState.shouldSyncPackets && (samplingState.updateRateLimiter.Allow() || samplingState.numCapturedPackets == samplingState.maxNumCapturedPackets)
+			if !shouldUpdate {
+				return nil
+			}
+		}
+
+		update := ps.DeepCopy()
+		if samplingState != nil {
+			update.Status.NumCapturedPackets = samplingState.numCapturedPackets
+		}
+
+		_, err = c.crdClient.CrdV1alpha1().PacketSamplings().UpdateStatus(context.TODO(), update, v1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("update Traceflow failed: %w", err)
+		}
+		klog.InfoS("Updated PacketSampling", "ps", klog.KObj(ps), "status", update.Status)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("PacketSampling update error: %w", err)
+	}
+
+	if samplingState != nil {
+		rawData := pktIn.Data.(*util.Buffer).Bytes()
+		ci := gopacket.CaptureInfo{
+			Timestamp:     time.Now(),
+			CaptureLength: len(rawData),
+			Length:        len(rawData),
+		}
+		err = samplingState.pcapngWriter.WritePacket(ci, rawData)
+		if err != nil {
+			return fmt.Errorf("couldn't write packet: %w", err)
+		}
+
+		if samplingState.numCapturedPackets == oldPS.Spec.FirstNSamplingConfig.Number && samplingState.shouldSyncPackets {
+			return c.uploadPacketsFile(oldPS)
+		}
+	}
+	return nil
+}
+
+// parsePacketIn parses the packet-in message and returns
+// 1. the sampling state of the PacketSampling (on sampling mode),
+func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (_ *crdv1alpha1.PacketSampling, _ *packetSamplingState, shouldSkip bool, _ error) {
+	// Get data plane tag.
+	// Directly read data plane tag from packet.
```

Comment:
I wonder in packet sampling case, do we really need to use the data plane tag? I think we don't inject any packets and we don't need a globally unique number to associate packets,right? Then can we just use a local regular reg to differentiate the packets?

Reply:
it's possible.  but like traceflow, it's not working in inter-node cases in noencap mode and  accoding to @gran-vmv , it's has known bugs in encap mode. I thought this would be the best choice for now.

Reply:
My understanding is we need the tag to identify the special generated packet in Traceflow case. Why does it also need the tag for PacketSampling where all packets are live traffic? Do I misunderstand the use case?

Reply:
Updated to use the last 4 bit in reg4 for flow match. I may have misunderstand the usercases for nonencap mode + inter-node, packetsampling doesn't affected by this(only need to capture at source or destionation location).